{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import plotly.express as px\n",
    "import plotly.graph_objects as go\n",
    "from scipy.stats import wilcoxon\n",
    "import numpy as np\n",
    "\n",
    "# supprime les avertissements de copies de dataframe\n",
    "# TODO : comprendre et supprimer correctement les avertissements\n",
    "pd.options.mode.chained_assignment = None\n",
    "\n",
    "IN = \"../../results/iadhore/all/\"\n",
    "OUT = \"../../results/python/\"\n",
    "WINDOW_SIZE = 90 # nombre de gènes dans la fenêtre glissante\n",
    "ANCHORS_MIN = 800\n",
    "ALPHA = 0.05\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Lecture et traitement de la table des multiplicons :\n",
    "id\tgenome_x\tlist_x\tparent\tgenome_y\tlist_y\tlevel\tnumber_of_anchorpoints\tprofile_length\tbegin_x\tend_x\tbegin_y\tend_y\tis_redundant\n",
    "\"\"\"\n",
    "\n",
    "df_multiplicons = pd.read_csv(IN + \"multiplicons.txt\", sep='\\t', index_col='id')\n",
    "print(df_multiplicons[:3], \"\\nlen:\", len(df_multiplicons))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Affichage du tableau du nbr de gènes homologues entre chaque chromosome. \"\"\"\n",
    "\n",
    "tmp = pd.concat( [ df_multiplicons['list_x'], df_multiplicons['list_y'] ] )\n",
    "chromosomes = list( pd.unique(tmp)) \n",
    "chromosomes = sorted(chromosomes)\n",
    "print(chromosomes, \"\\n\")\n",
    "\n",
    "table_nb_anchors = [[0 for x in range(len(chromosomes))] for x in range(len(chromosomes))]  # table de 0 de taille n*n \n",
    "#print(table_nb_anchors)\n",
    "\n",
    "# remplissage du nombre de points d'ancres entre chaque chromosome\n",
    "for i in range(len(chromosomes)) :\n",
    "    for j in range(i, len(chromosomes)) :\n",
    "        chr1 = chromosomes[i]\n",
    "        chr2 = chromosomes[j]\n",
    "        \n",
    "        rows = df_multiplicons[ ((df_multiplicons.list_x == chr1) & (df_multiplicons.list_y == chr2)) |\n",
    "            ((df_multiplicons.list_x == chr2) & (df_multiplicons.list_y == chr1)) ]\n",
    "        #print(rows[['list_x', 'list_y', 'number_of_anchorpoints']])\n",
    "        anchors = rows['number_of_anchorpoints']\n",
    "        total = anchors.values.sum()\n",
    "\n",
    "        table_nb_anchors[i][j] = total\n",
    "        table_nb_anchors[j][i] = total\n",
    "\n",
    "#print((table_nb_anchors))\n",
    "\n",
    "\"\"\"Affichage de la heat map du nombre de gènes semblables par paire de chromosomes\"\"\"\n",
    "heat = go.Heatmap(z=table_nb_anchors,\n",
    "                x=chromosomes,\n",
    "                y=chromosomes,\n",
    "                xgap=2, ygap=2\n",
    "                )\n",
    "layout = go.Layout(title_text=\"Heat map of anchor points in chromosomes\", title_x=0.5, \n",
    "                width=600, height=600,\n",
    "                xaxis_showgrid=False,\n",
    "                yaxis_showgrid=False,\n",
    "                yaxis_autorange='reversed'\n",
    "                )\n",
    "   \n",
    "fig=go.Figure(data=[heat], layout=layout)   \n",
    "fig.update_xaxes(side=\"top\")\n",
    "\n",
    "fig.write_html(OUT + \"anchors_chr_pair.html\")\n",
    "fig.show()\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Creation de la df des triplets de chromosomes dont le nbr d'ancres est superieur à un seuil\"\"\"\n",
    "MDchr = chromosomes[:17]\n",
    "PPchr = chromosomes[17:]\n",
    "print(MDchr, \"\\n\", PPchr, sep='')\n",
    "\n",
    "df_triplets = pd.DataFrame(columns=['PP', 'MD1', 'MD2', 'anchorpoints_1', 'anchorpoints_2'])\n",
    "\n",
    "\n",
    "\"\"\"Trouve le 3e chr d'un triplet avec les indices des 2 premiers, en respectant le seuil minimum d'ancres avec PP\"\"\"\n",
    "def third_chr(table_nb_anchors, chromosomes, iPP, iMD1) :\n",
    "    # MD2 est le chr dont l'indice est celui de la valeur max dans la table des ancres, dans la ligne de MD1, excluant les PP\n",
    "    max_MD1_MD2 = max(table_nb_anchors[iMD1][:17])\n",
    "    iMD2 = table_nb_anchors[iMD1].index(max_MD1_MD2)\n",
    "    MD2 = chromosomes[iMD2]\n",
    "\n",
    "    if table_nb_anchors[iMD2][iPP] > ANCHORS_MIN : return MD2, iMD2\n",
    "    else : return None, None\n",
    "\n",
    "\"\"\"Remplissage de la df grâce à la table du nombre d'ancres (gènes dupliqués)\"\"\"\n",
    "def remplissage_df_triplets(table_nb_anchors, chromosomes, df) :\n",
    "    for i in range(len(chromosomes)) :\n",
    "        for j in range(i, len(chromosomes)) :\n",
    "\n",
    "            # on conserve uniquement les matchs entre 2 espèces différentes, et supérieur au seuil \n",
    "            if table_nb_anchors[i][j] > ANCHORS_MIN and chromosomes[i][:2] != chromosomes[j][:2] :\n",
    "                PP = chromosomes[i] if chromosomes[i][:2] == \"Pp\" else chromosomes[j]\n",
    "                MD1 = chromosomes[i] if chromosomes[i][:2] == \"Md\" else chromosomes[j]\n",
    "                iMD1 = i if chromosomes[i][:2] == \"Md\" else j # l'indice du chromosome MD1\n",
    "                iPP = i + j - iMD1 # l'indice du chromosome PP\n",
    "                \n",
    "                # on trouve le 3e chromosome du triplet\n",
    "                MD2, iMD2 = third_chr(table_nb_anchors, chromosomes, iPP, iMD1)\n",
    "                if MD2 != None : \n",
    "                    MD = [MD1, MD2]\n",
    "                    MD.sort() # trie MD1 et MD2 par ordre alphabetique, pour simplifier la suppression des triplets en doublons après\n",
    "\n",
    "                    # crée la df d'une ligne et l'ajoute à df_triplets\n",
    "                    row = pd.DataFrame( {'PP':[PP], 'MD1':[MD[0]], 'MD2':[MD[1]], 'anchorpoints_1':[table_nb_anchors[iPP][iMD1]], 'anchorpoints_2':[table_nb_anchors[iPP][iMD2]]} )\n",
    "                    df = pd.concat( [df, row], ignore_index=True, axis=0 )\n",
    "\n",
    "    df.drop_duplicates(subset=['PP', 'MD1', 'MD2'], inplace=True, ignore_index=True) # suppression des triplets en double\n",
    "    df.sort_values('PP', inplace=True, ignore_index=True) # tri par valeurs de chromosome PP\n",
    "    return df\n",
    "\n",
    "\n",
    "df_triplets = remplissage_df_triplets(table_nb_anchors, chromosomes, df_triplets)\n",
    "print(df_triplets)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Lecture et traitement de la table des multiplicon_pairs :\n",
    "id\tmultiplicon\t\tgene_x\tgene_y\tchr_x chr_y\n",
    "\"\"\"\n",
    "\n",
    "df_pairs = pd.read_csv(IN + \"multiplicon_pairs_modified.txt\", sep='\\t', index_col='id')\n",
    "df_pairs.drop(['code'], axis=1, inplace=True) # supprime la colonne 'code'\n",
    "\n",
    "# jointure pour avoir les noms de chromosome de chaque gene\n",
    "df_pairs_chr = df_pairs.join(df_multiplicons[['list_x', 'list_y']], on='multiplicon')\n",
    "df_pairs_chr.rename(columns={'list_x':'chr_x', 'list_y':'chr_y'}, inplace=True)\n",
    "print(df_pairs_chr[1000:1010])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# construction de la liste des triplets de genes chez les chromosomes du triplet\n",
    "def make_df_genes_triplet(PP, MD1, MD2, df_pairs_chr) :\n",
    "    \n",
    "    # récupération les paires de gènes impliquant au moins un gène de PP\n",
    "    df_PPx = df_pairs_chr[ (df_pairs_chr.chr_x == PP) ] # le gène PP est dans le colonne chr_x\n",
    "    df_PPy = df_pairs_chr[ (df_pairs_chr.chr_y == PP) ] # le gène PP est dans le colonne chr_y\n",
    "    # échange de nom 2 colonnes d'une des deux df pour que tous les gènes du chr PP soient dans la meme colonne y\n",
    "    df_PPx.rename(columns={'gene_x':'gene_y', 'gene_y':'gene_x', 'chr_x':'chr_y', 'chr_y':'chr_x'}, inplace=True)\n",
    "    # concaténation sur la colonne y (le gène PP) des deux df \n",
    "    df_tmp = pd.concat([df_PPy, df_PPx]) \n",
    "\n",
    "    # sélection des paires dont le MD correspond à l'un des 2 chr donnés en arguments \n",
    "    df_MD1 = df_tmp[ (df_tmp.chr_x == MD1) ]\n",
    "    df_MD2 = df_tmp[ (df_tmp.chr_x == MD2) ]\n",
    "\n",
    "    # merge des 2 df, c'est-a-dire réalise une jointure inclusive\n",
    "    df_triplet = pd.merge(df_MD1, df_MD2, on=['gene_y', 'chr_y'], how='outer', indicator=True)\n",
    "    df_triplet = df_triplet.filter({'gene_y', 'gene_x_x', 'gene_x_y'}) # conservation des colonnes des gènes PP MD1 et MD2\n",
    "    df_triplet.rename(columns={'gene_y':'gene_PP', 'gene_x_x':'gene_MD1', 'gene_x_y':'gene_MD2'}, inplace=True) # renomme les colonnes\n",
    "    df_triplet = df_triplet.reindex(columns=['gene_PP', 'gene_MD1', 'gene_MD2']) # réordonne les colonnes\n",
    "    df_triplet.sort_values('gene_PP', inplace=True, ignore_index=True) # trie les lignes par PP croissant, ordonné comme sur le chromosome\n",
    "\n",
    "    #df_triplet.to_csv(OUT + \"tmp_\" + PP + MD1 + MD2 + \".csv\", index=False)\n",
    "\n",
    "    return df_triplet\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Création d'une df mesurant le nombre de gènes conservés à chaque itération de la fenêtre le long du chromosome :\n",
    "\n",
    "iteration    gene_PP          gene_MD1     nb_MD1 rate_MD1    gene_MD2     nb_MD2 rate_MD2\n",
    "1            Prupe.3G011400.1 MD09G1282900 72.0   80.000000   MD17G1272600 73.0   81.111111  \n",
    "2            Prupe.3G011500.1 MD09G1282700 73.0   81.111111   MD17G1272500 73.0   81.111111  \n",
    "3            Prupe.3G011600.1 MD09G1282600 74.0   82.222222   MD17G1272400 73.0   81.111111  \n",
    "4            Prupe.3G011700.1          NaN 73.0   81.111111   MD17G1272300 73.0   81.111111  \n",
    "\"\"\"\n",
    "\n",
    "def make_df_fractionation(df_genes) :\n",
    "    # copie les colonnes des genes MD1 et MD2 dans la new df\n",
    "    df_fractionation = pd.DataFrame(dtype=float)\n",
    "    df_fractionation['nb_MD1'] = df_genes['gene_MD1']\n",
    "    df_fractionation['nb_MD2'] = df_genes['gene_MD2']\n",
    "\n",
    "    # remplace tous les noms de genes par la valeur 1 et maintient les NaN tel quel\n",
    "    df_fractionation = df_fractionation.applymap(lambda x: np.nan if pd.isna(x) else 1)\n",
    "\n",
    "    # compte le nb de genes de MD1 dans la fenetre\n",
    "    df_fractionation.nb_MD1 = df_fractionation.nb_MD1.rolling(WINDOW_SIZE, min_periods=0).count()\n",
    "    # compte le nb de genes de MD2 dans la fenetre\n",
    "    df_fractionation.nb_MD2 = df_fractionation.nb_MD2.rolling(WINDOW_SIZE, min_periods=0).count()\n",
    "    \n",
    "    # copies les colonnes de df_genes avant modification \n",
    "    df_fractionation = pd.concat([df_fractionation, df_genes[['gene_PP', 'gene_MD1', 'gene_MD2']]], axis=1)\n",
    "\n",
    "    # calcule les pourcentages de conservation des gènes en fonction de la taille de la fenetre\n",
    "    df_fractionation['rate_MD1'] = df_fractionation['nb_MD1'] * 100 / WINDOW_SIZE\n",
    "    df_fractionation['rate_MD2'] = df_fractionation['nb_MD2'] * 100 / WINDOW_SIZE\n",
    "\n",
    "    # ajoute colonne du numero d'itération de la fenêtre, +1 normal et +1 pour que itération commence à 1 au lieu de 0 \n",
    "    df_fractionation['iteration'] = df_fractionation.index - WINDOW_SIZE + 2\n",
    "    # réordonne les colonnes\n",
    "    df_fractionation = df_fractionation.reindex(columns=['iteration', 'gene_PP', 'gene_MD1', 'nb_MD1', 'rate_MD1', 'gene_MD2', 'nb_MD2', 'rate_MD2']) \n",
    "    # remplace par NaN les valeurs de conservation de genes pour les fenêtres du début du chromosome, plus petites que la WINDOW_SIZE\n",
    "    df_fractionation.loc[:WINDOW_SIZE-2, ['rate_MD1', 'nb_MD1', 'rate_MD2', 'nb_MD2']] = np.nan \n",
    "    df_fractionation.loc[:WINDOW_SIZE-2, ['iteration']] = 0 # remplace par 0 les iterations des fenêtres du début\n",
    "\n",
    "    # converti les valeurs de la colonne iteration en entiers\n",
    "    df_fractionation = df_fractionation.astype({'iteration': int})    \n",
    "    \n",
    "    return df_fractionation\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Affiche le graphique du taux de conservation de genes au sein de 2 chromosomes dupliqués\"\"\"\n",
    "def display_graph_fractionation(df_display, triplet) :\n",
    "    MD1 = triplet.get(\"MD1\")\n",
    "    MD2 = triplet.get(\"MD2\")\n",
    "    PP = triplet.get(\"PP\")\n",
    "\n",
    "    # renomme les colonnes pour avoir une légende compréhensible\n",
    "    df = df_display.rename(columns={'rate_MD1': MD1, 'rate_MD2': MD2})\n",
    "    #print(df)\n",
    "\n",
    "    # diagramme à ligne brisée du pourcentage de conservation des gènes le long de 2 chromosomes\n",
    "    fig = px.line(df, x='iteration', y=[MD1, MD2])\n",
    "    fig.update_layout(xaxis_title=\"window (size = \" + str(WINDOW_SIZE) + \") iteration along \" + PP,\n",
    "                    yaxis_title=\"genome conservation rate (%)\",\n",
    "                    title=\"Fractionation biais between \" + MD1 + \" and \" + MD2,\n",
    "                    xaxis_range=[0, len(df)],\n",
    "                    yaxis_range=[-1, 101] )\n",
    "    \n",
    "    fig.write_html(OUT + MD1 + \"_\" + MD2 + \"_\" + PP + \".html\")\n",
    "    fig.show()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Parcourt la liste des triplets de chromosomes pour en faire des graphes de biais de fractionnement\"\"\"\n",
    "def analysis_each_triplet(df_triplets) :\n",
    "    for triplet in df_triplets.to_dict('records') :\n",
    "       analysis_one_triplet(triplet)\n",
    "\n",
    "\"\"\"Graphe le gene fractionation d'un triplet et réalise son test statistique\"\"\"\n",
    "def analysis_one_triplet(triplet) :\n",
    "    # traitement des données\n",
    "    df_genes_triplet = pd.DataFrame()\n",
    "    df_genes_triplet = make_df_genes_triplet(triplet.get('PP'), triplet.get('MD1'), triplet.get('MD2'), df_pairs_chr)\n",
    "    df_fractionation = make_df_fractionation(df_genes_triplet)\n",
    "    # suppression des NaN des données de pourcentage de conservation des gènes\n",
    "    df_display = df_fractionation.dropna(subset=['rate_MD1', 'rate_MD2'])\n",
    "\n",
    "    # affichage des données\n",
    "    print(triplet)\n",
    "    #print(df_display)\n",
    "    interpretation_test(df_display)\n",
    "    display_graph_fractionation(df_display, triplet)\n",
    "\n",
    "\"\"\"Réalise et interprete le test statistique wilcoxons\"\"\"\n",
    "def interpretation_test(df_display) :\n",
    "    res = wilcoxon(df_display['rate_MD1'], df_display['rate_MD2'])\n",
    "    print(res)\n",
    "    if (res.pvalue < ALPHA) : print(\"TEST SIGNIFICATIF : il existe un biais de fractionnement au risque alpha=\", ALPHA, \". \", sep='')\n",
    "    else : print(\"TEST NON SIGNIFICATIF : il n'existe pas de biais de fractionnement au risque alpha=\", ALPHA, \". \", sep='')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"test pour le premier triplet de la df contenant PP comme chromosome\"\"\"\n",
    "def test(df_triplets, PP) :\n",
    "    triplet = (df_triplets[ df_triplets.PP == PP ]).to_dict('records')[0]\n",
    "    analysis_one_triplet(triplet)\n",
    "\n",
    "\n",
    "\"\"\"MAIN\"\"\"\n",
    "if __name__==\"__main__\" :\n",
    "    \n",
    "    # lance l'analyse de tous les triplets trouvés selon le nbr de gènes similaires\n",
    "    #analysis_each_triplet(df_triplets)\n",
    "\n",
    "    # lance l'analyse d'un triplet\n",
    "    test(df_triplets, \"Pp03\")"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "31f2aee4e71d21fbe5cf8b01ff0e069b9275f58929596ceb00d14d90e3e16cd6"
  },
  "kernelspec": {
   "display_name": "Python 3.9.2 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
